### PR TITLE
Build Pip Wheels

### LINKED ISSUE #473: Pip installation fails, `CUTCLASS` not found
# 🐛 Bug

pip installation fails in a docker container, CUTCLASS not found,  `git submodule update --init --recursive` not executed

## To Reproduce

Dockerfile
```Dockerfile
FROM pytorch/pytorch:1.12.1-cuda11.3-cudnn8-runtime
RUN pip install xformers
```
then
```
docker build .
```
## Error Trace
<details>
  <summary>open</summary>

```
#1 [internal] load build definition from Dockerfile
#1 sha256:bc3772a9760c6470030d3506e7afa0b9caa2a77f63376fe30fc296a334d5c980
#1 transferring dockerfile: 116B done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 sha256:5b674e66e988c8852edbf605c0d0921ac6eed40841cd55d9112e0d92242091a1
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/pytorch/pytorch:1.12.1-cuda11.3-cudnn8-runtime
#3 sha256:409f78a4f3551ef4b6d7a4b064ff72bb54f0677d599351b4d0dcdff08b926834
#3 DONE 0.8s

#4 [1/2] FROM docker.io/pytorch/pytorch:1.12.1-cuda11.3-cudnn8-runtime@sha256:0bc0971dc8ae319af610d493aced87df46255c9508a8b9e9bc365f11a56e7b75
#4 sha256:2e3e89abd93f2e7b42b070196f0e6be4ce38a2d360c98232440e1d90189bdb02
#4 CACHED

#5 [2/2] RUN pip install xformers
#5 sha256:ef3133015f56a22d509f2aa1ef730afdcaa2591838105ba332650ff73ceb9ff9
#5 1.012 Collecting xformers
#5 1.313   Downloading xformers-0.0.13.tar.gz (292 kB)
#5 1.429      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 292.5/292.5 kB 2.6 MB/s eta 0:00:00
#5 1.534   Preparing metadata (setup.py): started
#5 2.952   Preparing metadata (setup.py): finished with status 'error'
#5 2.961   error: subprocess-exited-with-error
#5 2.961
#5 2.961   × python setup.py egg_info did not run successfully.
#5 2.961   │ exit code: 1
#5 2.961   ╰─> [8 lines of output]
#5 2.961       Traceback (most recent call last):
#5 2.961         File "<string>", line 36, in <module>
#5 2.961         File "<pip-setuptools-caller>", line 34, in <module>
#5 2.961         File "/tmp/pip-install-94ty405p/xformers_31debcecca1f46019eadae6eead5cc3f/setup.py", line 239, in <module>
#5 2.961           ext_modules=get_extensions(),
#5 2.961         File "/tmp/pip-install-94ty405p/xformers_31debcecca1f46019eadae6eead5cc3f/setup.py", line 158, in get_extensions
#5 2.961           "CUTLASS submodule not found. Did you forget "
#5 2.961       RuntimeError: CUTLASS submodule not found. Did you forget to run `git submodule update --init --recursive` ?
#5 2.961       [end of output]
#5 2.961
#5 2.961   note: This error originates from a subprocess, and is likely not a problem with pip.
#5 2.965 error: metadata-generation-failed
#5 2.965
#5 2.965 × Encountered error while generating package metadata.
#5 2.965 ╰─> See above for output.
#5 2.965
#5 2.965 note: This is an issue with the package mentioned above, not pip.
#5 2.965 hint: See above for details.
#5 ERROR: executor failed running [/bin/sh -c pip install xformers]: exit code: 1
------
 > [2/2] RUN pip install xformers:
------
executor failed running [/bin/sh -c pip install xformers]: exit code: 1
```
</details>

## Expected behavior

installation should work.

## Environment

in the container, running docker on windows

<details>
  <summary>open</summary>

```
PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A 

OS: Ubuntu 18.04.6 LTS (x86_64) 
GCC version: Could not collect  
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.17

Python version: 3.7.13 (default, Mar 29 2022, 02:18:16)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.10.16.3-microsoft-standard-WSL2-x86_64-with-debian-buster-sid
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1060
Nvidia driver version: 517.48
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.5
[pip3] torch==1.12.1
[pip3] torchtext==0.13.1
[pip3] torchvision==0.13.1
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.3.1               ha36c431_9    nvidia
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl-service               2.4.0            py37h7f8727e_0
[conda] mkl_fft                   1.3.1            py37hd3c417c_0
[conda] mkl_random                1.2.2            py37h51133e4_0
[conda] numpy                     1.21.5           py37he7a7128_2
[conda] numpy-base                1.21.5           py37hf524024_2
[conda] pytorch                   1.12.1          py3.7_cuda11.3_cudnn8.3.2_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchtext                 0.13.1                     py37    pytorch
[conda] torchvision               0.13.1               py37_cu113    pytorch
```

</details>

## Additional context

I don't think this problem has anything to do with os/python/pytorch/cuda/nvcc versions, the `setup.py` seems to be tailored for local / manual install, and fails in the context.

### PR DESCRIPTION
**Pull request by AbdBarho**
## What does this PR do?
Adds Github Workflow to build pip wheels. Fixes #473.

## Before submitting

- [ ] Did you have fun?
  - changing one line and then waiting a couple of hours to find out that I made a typo is not what I would define as fun, but I definitely learned a lot which is why I wanted to do this in the first place.
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/xformers/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - #473
- [ ] Did you make sure to update the docs?
  - The xformers team should decide on the release and delivery processes of the pip wheels first, then the docs can be updated.
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/xformers/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.

### CODE REVIEW

**Comment by AbdBarho** [comment]
OOooops sorry! opened an MR in this repo instead of mine!

**Comment by danthe3rd** [comment]
Oh btw we would love to have official wheels in the repo, if you want to contribute a PR :)

**Comment by AbdBarho** [comment]
@danthe3rd  Thanks for following up after my mistake! 

I can gladly re-open this MR, it was intended as part of the discussion in https://github.com/facebookresearch/xformers/issues/473#issuecomment-1280026134.

However, there a lot of stuff that is still unclear to me, maybe you or anyone else on the team can give me more input on this, just so I can know what to focus on:
* The build is done in github workflows, but I see your ci infra is more towards circle ci, would this be a problem?
  * I am using github workflows basically as "free cloud compute" since my 2015 laptop would probably melt before finishing installing `nvcc`
* There is no automatic publishing of the wheels, instead, you get an `artefacts.zip` file at the end of the workflow which has to be downloaded / extracted / published manually (although this should be easy to fix, assuming the secrets are available)
* I am currently building a wheel for each different python version, which seems like an absolute overkill, especially since the core of this repo is not python heavy, I am no `pip`/`build` expert, is there a way to build for all python versions as one package?
  * Each wheel is roughly 200MB.
* The build takes at least 1.5 hours per version (single threaded), with more CUDA architectures it takes even more, using ninja will get the runner killed for hugging resources. I am unsure if this can be optimized.
* I have no idea of how to build / compile this repo on windows, I am also not sure if github provides windows runners in the non-enterprise version.
* what to do with the "old" build scripts / infra ? would you prefer a github workflows only solution? or something more [scripty](https://github.com/facebookresearch/xformers/blob/836768506d990fa3985dafd69a0b4518b3ae12bd/packaging/build_wheel.sh)?

Those are some of the questions that I have, the rest is just a consequence of the above.

Thank you! have a nice weekend.

**Comment by danthe3rd** [comment]
Let me get back to you next week. @fmassa @blefaudeux and @bottler might have some opinion as well

**Comment by blefaudeux** [comment]
> Let me get back to you next week. @fmassa @blefaudeux and @bottler might have some opinion as well

pulling in @patrickvonplaten and @patil-suraj also, who were interested in helping or looking at that wheel building procedure at least

**Comment by blefaudeux** [comment]
Not personally in charge anymore, but trying to help out still :)

> @danthe3rd Thanks for following up after my mistake!
> 
> I can gladly re-open this MR, it was intended as part of the discussion in [#473 (comment)](https://github.com/facebookresearch/xformers/issues/473#issuecomment-1280026134).
> 
> However, there is a lot of stuff that is still unclear to me, maybe you or anyone else on the team can give me more input on this, just so I can know what to focus on:
> 
>     * The build is done in github workflows, but I see your ci infra is more towards circle ci, would this be a problem?
       
I'm actually amazed that this works out on GHA, I didn´t think that was possible. I'm guessing that it will quickly count against the Meta credits, but it could be tied to a release (vs. for all PRs) and that would be probably good enough ?

>       * I am using github workflows basically as "free cloud compute" since my 2015 laptop would probably melt before finishing installing `nvcc`

ahah, totally understood ! 

> 
>     * There is no automatic publishing of the wheels, instead, you get an `artefacts.zip` file at the end of the workflow which has to be downloaded / extracted / published manually (although this should be easy to fix, assuming the secrets are available)
> 

This can be automated indeed I believe, as you mentioned. I actually think that fairseq already has[ the flows ](https://github.com/facebookresearch/fairseq/blob/main/.github/workflows/release.yml) for that

>     * I am currently building a wheel for each different python version, which seems like an absolute overkill, especially since the core of this repo is not python heavy, I am no `pip`/`build` expert, is there a way to build for all python versions as one package?
> 
Not a real expert also, but should not be a blocker I think, these days it's trivial to change python versions (and most of them should be ABI compatible but I'm not 100% sure there)
      
>       * Each wheel is roughly 200MB.
> 
>     * The build takes at least 1.5 hours per version (single threaded), with more CUDA architectures it takes even more, using ninja will get the runner killed for hugging resources. I am unsure if this can be optimized.
> 
>     * I have no idea of how to build / compile this repo on windows, I am also not sure if github provides windows runners in the non-enterprise version.

Not sure either but same as above, should not be a blocker I think, let's not make perfect the enemy of great :)

>     * what to do with the "old" build scripts / infra ? would you prefer a github workflows only solution? or something more [scripty](https://github.com/facebookresearch/xformers/blob/836768506d990fa3985dafd69a0b4518b3ae12bd/packaging/build_wheel.sh)?

as far as I'm concerned anything old can get the door if anything new is better :dancers: 

> Those are some of the questions that I have, the rest is just a consequence of the above.
> 
> Thank you! have a nice weekend.

Thank you, I didn´t know that this was up but it's great work, could be super duper useful to many people

File: .github/workflows/wheels.yml
Comment (danthe3rd) [inline .github/workflows/wheels.yml:14]: Can we also add `8.0` here for A100?

File: .github/workflows/wheels.yml
Comment (danthe3rd) [inline .github/workflows/wheels.yml:21]: let's support python `3.9` and `3.10` to be coherent with the [conda builds](https://anaconda.org/xformers/xformers/files)

File: .github/workflows/wheels.yml
Comment (danthe3rd) [inline .github/workflows/wheels.yml:39]: we should install the version corresponding to `matrix.cuda`. There are some available github-action for that to help apparently: https://github.com/Jimver/cuda-toolkit

File: .github/workflows/wheels.yml
Comment (danthe3rd) [inline .github/workflows/wheels.yml:49]: > using ninja will get the runner killed for hugging resources

This can be fixed by setting `MAX_JOBS=2` for instance, but I'm not even sure this would be enough - as the github runners [only have 7GB of RAM](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources)

**Review by danthe3rd — COMMENTED** [review]
Actually the code in the PR looks good for a v1 - we can take care of the upload+windows later. Just added a few comments before we can merge it

File: .github/workflows/wheels.yml
Comment (patrickvonplaten) [inline .github/workflows/wheels.yml:21]: Is pip currently not available for 3.7, 3.8, 3.9?

**Comment by patrickvonplaten** [comment]
Hey @AbdBarho,

It's extremely cool that you're working on making pip wheels easily available for the community! I think many people use xformers fast attention mechanism and would love an easy way of installing it via pip.

As maintainers of the diffusers library: https://github.com/huggingface/diffusers we'd love to more prominently feature xformers for training and would be very happy to help you! We're also in contact with the PyTorch team (gently pinging @ngimel just FYI). 

Please let us know if you're stuck with anything (github actions etc...) => we're happy to help!

File: .github/workflows/wheels.yml
Comment (AbdBarho) [reply .github/workflows/wheels.yml]: ok!

File: .github/workflows/wheels.yml
Comment (AbdBarho) [reply .github/workflows/wheels.yml]: I am going to have to try this first.

Currenty, the build happens in a debian slim container, which should (at least in theory) work with this action.

PyTorch logs a warning to the console if the nvcc version is not the same as the one it was compiled with, but because it is only a minor version, pytorch logs that "this is likely not a problem" so I just winged it.

File: .github/workflows/wheels.yml
Comment (AbdBarho) [reply .github/workflows/wheels.yml]: Yeah, I tried with 2, it still got killed.

File: .github/workflows/wheels.yml
Comment (AbdBarho) [reply .github/workflows/wheels.yml:21]: Quite the opposite, it is possible to compile to all of the above versions, this was just a test

**Comment by AbdBarho** [comment]
@patrickvonplaten Thank you very much!

I am going to try to fit in this in my schedule tomorrow, I will report back if I face any problems.

**Comment by AbdBarho** [comment]
This PR is now ready for review.

You can see the workflow and download the generated wheels here:
https://github.com/AbdBarho/xformers-wheels/actions/runs/3499213686

I ask you all please to test the built wheels in your applications, I did a test with python3.10, torch1.12.1, and cuda 11.6 on [my SD docker](https://github.com/AbdBarho/stable-diffusion-webui-docker), and it worked without a problem, the more tests the better.

File: .github/workflows/wheels.yml
Comment (AbdBarho) [reply .github/workflows/wheels.yml]: I ended up not using this action since it only works on debian but the container I am using is centos

**Review by danthe3rd — APPROVED** [review]
This looks great thank you! The CI failing is not due to this PR, so I'll merge.
Thanks a lot for your contribution!

@patrickvonplaten if you want to improve this workload, feel free to open a new PR.
In list of remaining things to do, I think we need to worry about windows as well (and as an extension, it would be great to have a CI for windows as well on all pull requests).

We don't plan to contribute to it in the short term, the only thing we will do is adding upload to [pypi](https://pypi.org/project/xformers/)

**Comment by susumuota** [comment]
@AbdBarho 
Hi, thanks you for the great PR!

> I ask you all please to test the built wheels in your applications, I did a test with python3.10, torch1.12.1, and cuda 11.6 on [my SD docker](https://github.com/AbdBarho/stable-diffusion-webui-docker), and it worked without a problem, the more tests the better.

I have tested 3.10.6, 3.7.12 on GCE with Deep Learning VM and 3.7.15 at Google Colab.
There was no problem on 3.10.6 but I found a minor bug on 3.7s. Also, I fixed it and tested it on 3.7.12, 3.7.15 and 3.10.6.
I attached a Colab Notebook for details.

@danthe3rd 
I got SyntaxError at `/usr/local/lib/python3.7/dist-packages/xformers/ops/unbind.py`, line 99

```
Python 3.7.15 (default, Oct 12 2022, 19:14:55) 
[GCC 7.5.0]
Commit hash: 47a44c7e421b98ca07e92dbf88769b04c9e28f86
Installing requirements for Web UI
Launching Web UI with arguments: --xformers --share
Cannot import xformers
Traceback (most recent call last):
  File "/content/stable-diffusion-webui/modules/sd_hijack_optimizations.py", line 18, in <module>
    import xformers.ops
  File "/usr/local/lib/python3.7/dist-packages/xformers/ops/__init__.py", line 19, in <module>
    from .swiglu_op import (  # noqa: F401
  File "/usr/local/lib/python3.7/dist-packages/xformers/ops/swiglu_op.py", line 14, in <module>
    from .unbind import stack_or_none, unbind
  File "/usr/local/lib/python3.7/dist-packages/xformers/ops/unbind.py", line 99
    return None, *grad.unbind(dim=ctx.dim)
                 ^
SyntaxError: invalid syntax
```

I tried to find info on xformers repository.

```
        return None, *grad.unbind(dim=ctx.dim)
```

Then I found this [commit (#527)](https://github.com/facebookresearch/xformers/commit/b8e686fa538e30553bace9d39188b3cb83d34284) by @danthe3rd , and tried to fix like this,

```
        return None, *(grad.unbind(dim=ctx.dim))
```

But still it didn't work. Then, I found a info on [Python 3.8's what's new page](https://docs.python.org/3/whatsnew/3.8.html#other-language-changes).

> Generalized iterable unpacking in yield and return statements no longer requires enclosing parentheses.
so in Python 3.7, it requires enclosing parentheses at return statements!

It's a matter of return statement. So, this 

```
        return None, *grad.unbind(dim=ctx.dim)
```

should be written like this in Python 3.7.

```
        return (None, *grad.unbind(dim=ctx.dim))
```

Here is a point of the difference between Python 3.7 and 3.8+.

```
$ python3.10
Python 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:36:39) [GCC 10.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> def test1():
...     a = (1, 2, 3)
...     return None, *a
...
>>> test1()
(None, 1, 2, 3)
>>> def test2():
...     a = (1, 2, 3)
...     return None, *(a)
...
>>> test2()
(None, 1, 2, 3)
>>> def test3():
...     a = (1, 2, 3)
...     return (None, *a)
...
>>> test3()
(None, 1, 2, 3)
```

```
$ python3.7
Python 3.7.12 | packaged by conda-forge | (default, Oct 26 2021, 06:08:21)
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> def test1():
...     a = (1, 2, 3)
...     return None, *a
  File "<stdin>", line 3
    return None, *a
                 ^
SyntaxError: invalid syntax
>>> def test2():
...     a = (1, 2, 3)
...     return None, *(a)
  File "<stdin>", line 3
    return None, *(a)
                 ^
SyntaxError: invalid syntax
>>> def test3():
...     a = (1, 2, 3)
...     return (None, *a)
...
>>> test3()
(None, 1, 2, 3)
```

Thanks!
[xformers_py37_whl_fix.ipynb.zip](https://github.com/facebookresearch/xformers/files/10050095/xformers_py37_whl_fix.ipynb.zip)

**Comment by patrickvonplaten** [comment]
Are pip wheels working now that this PR is merged?

**Pull request merged**
